File: src/lib/api/applications.ts
import { apiRequest } from '@/lib/api/client';
import type {
  Application,
  ApplicationCreatePayload,
  ApplicationSourceType,
  ApplicationStatus,
} from '@/lib/types';

type ApplicationDto = {
  id: string;
  source_type: ApplicationSourceType;
  source_id: string | null;
  source_title: string | null;
  full_name: string;
  email: string;
  phone: string;
  organization: string | null;
  title: string | null;
  message: string | null;
  status: ApplicationStatus;
  created_at: string;
};

function mapApplication(dto: ApplicationDto): Application {
  return {
    id: dto.id,
    sourceType: dto.source_type,
    sourceId: dto.source_id,
    sourceTitle: dto.source_title,
    fullName: dto.full_name,
    email: dto.email,
    phone: dto.phone,
    organization: dto.organization,
    title: dto.title,
    message: dto.message,
    status: dto.status,
    createdAt: new Date(dto.created_at),
  };
}

function mapCreatePayload(payload: ApplicationCreatePayload) {
  return {
    source_type: payload.sourceType,
    source_id: payload.sourceId ?? null,
    source_title: payload.sourceTitle ?? null,
    full_name: payload.fullName,
    email: payload.email,
    phone: payload.phone,
    organization: payload.organization ?? null,
    title: payload.title ?? null,
    message: payload.message ?? null,
  };
}

export const applicationsApi = {
  create: async (payload: ApplicationCreatePayload): Promise<Application> => {
    const data = await apiRequest<ApplicationDto>('/api/applications', {
      method: 'POST',
      body: JSON.stringify(mapCreatePayload(payload)),
    });
    return mapApplication(data);
  },

  listAdmin: async (params?: {
    sourceType?: ApplicationSourceType;
    sourceId?: string;
  }): Promise<Application[]> => {
    const search = new URLSearchParams();
    if (params?.sourceType) search.set('source_type', params.sourceType);
    if (params?.sourceId) search.set('source_id', params.sourceId);
    const qs = search.toString();
    const data = await apiRequest<ApplicationDto[]>(
      `/api/admin/applications${qs ? `?${qs}` : ''}`
    );
    return data.map(mapApplication);
  },

  updateStatus: async (id: string, status: ApplicationStatus): Promise<Application> => {
    const data = await apiRequest<ApplicationDto>(`/api/admin/applications/${id}`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    });
    return mapApplication(data);
  },
};

File: backend/app/routes/applications.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_admin, get_db
from app.models.application import Application
from app.schemas.application import (
    ApplicationCreate,
    ApplicationOut,
    ApplicationSourceType,
    ApplicationStatus,
    ApplicationStatusUpdate,
)

public_router = APIRouter(prefix="/api/applications", tags=["applications"])
admin_router = APIRouter(prefix="/api/admin/applications", tags=["admin:applications"])


@public_router.post("", response_model=ApplicationOut, status_code=201)
def create_application(body: ApplicationCreate, db: Session = Depends(get_db)):
    obj = Application(
        source_type=body.source_type.value,
        source_id=body.source_id,
        source_title=body.source_title,
        full_name=body.full_name.strip(),
        email=body.email.strip().lower(),
        phone=body.phone.strip(),
        organization=body.organization,
        title=body.title,
        message=body.message,
        status=ApplicationStatus.new.value,
    )
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@admin_router.get("", response_model=list[ApplicationOut])
def list_applications(
    source_type: ApplicationSourceType | None = Query(default=None),
    source_id: UUID | None = Query(default=None),
    db: Session = Depends(get_db),
    _=Depends(get_current_admin),
):
    query = db.query(Application)
    if source_type is not None:
        query = query.filter(Application.source_type == source_type.value)
    if source_id is not None:
        query = query.filter(Application.source_id == source_id)
    return query.order_by(Application.created_at.desc()).all()


@admin_router.patch("/{application_id}", response_model=ApplicationOut)
def update_application_status(
    application_id: UUID,
    body: ApplicationStatusUpdate,
    db: Session = Depends(get_db),
    _=Depends(get_current_admin),
):
    obj = db.query(Application).filter(Application.id == application_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Application not found")
    obj.status = body.status.value
    db.commit()
    db.refresh(obj)
    return obj
